plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/TimelineTable.java

```diff
@@ -0,0 +1,137 @@
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.hudi.HudiUtil.buildTableMetaClient;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+
+public class TimelineTable
+        implements SystemTable
+{
+    private final ConnectorTableMetadata tableMetadata;
+    private final HoodieTableMetaClient metaClient;
+    private final List<Type> types;
+
+    public TimelineTable(Configuration configuration, SchemaTableName tableName, Table hudiTable)
+    {
+        this.tableMetadata = new ConnectorTableMetadata(requireNonNull(tableName, "tableName is null"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata("timestamp", VARCHAR))
+                        .add(new ColumnMetadata("action", VARCHAR))
+                        .add(new ColumnMetadata("state", VARCHAR))
+                        .build());
+        this.metaClient = buildTableMetaClient(configuration, hudiTable.getStorage().getLocation());
+        this.types = tableMetadata.getColumns().stream().map(ColumnMetadata::getType).collect(toImmutableList());
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return Distribution.SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        Iterator<HoodieInstant> hudiInstants = metaClient.getActiveTimeline().getInstants().iterator();
+        if (!hudiInstants.hasNext()) {
+            return InMemoryRecordSet.builder(types).build().cursor();
+        }
+        HudiInstantIterable hudiInstantIterable = new HudiInstantIterable(hudiInstants, types);
+        return hudiInstantIterable.cursor();
+    }
+
+    private static class HudiInstantIterable
+            implements Iterable<List<Object>>
+    {
+        private final Iterator<HoodieInstant> hudiInstants;
+        private final List<Type> types;
+
+        public HudiInstantIterable(Iterator<HoodieInstant> hudiInstants, List<Type> types)
+        {
+            this.hudiInstants = requireNonNull(hudiInstants, "hudiInstants is null");
+            this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        }
+
+        public RecordCursor cursor()
+        {
+            return new InMemoryRecordSet.InMemoryRecordCursor<Iterator<? extends List<?>>>(types, this.iterator())
+            {
+                @Override
+                public void close()
+                {
+                    records = null;
+                }
+            };
+        }
+
+        @Override
+        public Iterator<List<Object>> iterator()
```

Comment:
There's no need in your case to create an anonymous class.

See https://github.com/trinodb/trino/blob/23e357ede36114711d9446d248180ffd2a564248/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java